File: problem_059.py
#!/usr/bin/env python
import re

print("Project Euler - Problem 42:")
print("Decrypt the message and find the sum of the ASCII values in the original text.\n")

def isPlainASCII(c):
    if c > ord("z"): return False
    if c < ord(" "): return False
    return True

def seemsLegit(text):
    text = text.split(" ")
    for word in text:
        if len(word) > 15:
            return False
    return True

def findCipher(inBuffer):
    # Cipher is 3 letters, so brute force all combinations
    for A in range(ord('a'), ord('z')):
        for B in range(ord('a'), ord('z')):
            for C in range(ord('a'), ord('z')):
                cipher = [A, B, C]
                outBuffer = ""
                bufferSum = 0
                solution = True
                # Take 3 letters at a time
                for i in range(0, len(inBuffer), 3):
                    # For each letter, decipher, and see if its plain ascii
                    for j in range(0, 3):
                        if i+j < len(inBuffer):
                            char = int(inBuffer[i+j])^cipher[j]
                            if isPlainASCII(char) == False:
                                solution = False
                                break
                            outBuffer += chr(char)
                            bufferSum += char
                    if solution == False: break

                # If it is plain ascii, take whole paragraph and see if it looks ok-ish
                if solution == True and seemsLegit(outBuffer):
                    return chr(A)+chr(B)+chr(C), outBuffer, bufferSum

# Read file in, and put into a list
ReadFile = open("problem_059.txt")
inBuffer = ReadFile.readline()
inBuffer = inBuffer.split(",")

# Get the answer
cipher, text, asciiSum = findCipher(inBuffer)

print("Cipher: %s" % cipher)
print(text)
print("")
print(asciiSum)

File: problem_059.c
// -*- compile-command: "gcc -o problem_059 problem_059.c -Wall" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <iso646.h>
#include <stdbool.h>
#include <stdlib.h>

FILE *cipherFile;
const int maxCipherSize = 1500;

int decode(int message[], const int length)
{
    int decode[length];
    int key1, key2, key3;

    // Go through possible cipher combinations (3 word key)
    for (key1='a' ; key1<='z' ; key1++)
        for (key2='a' ; key2<='z' ; key2++)
            for (key3='a' ; key3<='z' ; key3++)
	    {
                bool skip = false;
                int index;
                // Decipher message with key
                for (index=0 ; index<length ; index+=3)
		{
                    decode[index+0] = (message[index+0] ^ key1);
                    decode[index+1] = (message[index+1] ^ key2);
                    decode[index+2] = (message[index+2] ^ key3);
                }

                // Filter results. Return first probable result.
                int wordLen = 0;
                for (index=0 ; index<length ; index++)
                {
                    //Check for weird characters
                    if (decode[index] < ' ' || decode[index] > '}')
                    {
                        skip = true;
                        break;
                    }

                    //Check for unreasonably long words
                    if (decode[index] == 32)
                        wordLen = 0;
                    else
                        wordLen++;
                    if (wordLen > 14)
                    {
                        skip = true;
                        break;
                    }
                }

                if (skip == true) continue;

                else // We have a result
                {
                    int sum = 0;
                    // Print deciphered message
                    printf("\nCipher: %c%c%c \n", key1, key2, key3);
                    for (index=0 ; index<length ; index++)
		    {
                        printf("%c", decode[index]);
                        sum+=decode[index];
		    }
                    printf("\n");

                    // Return sum
                    return sum;
                }
            }
    return 0;
}

int main()
{
     printf("Project Euler - Problem 59:\n"
            "Decrypt the message and find the sum of the ASCII values in the original text.\n\n");

    int message[maxCipherSize];

    // Read in cipher to array; Convert ascii to numbers
    cipherFile = fopen ("problem_059.txt", "rt");
    char ch;
    int index = 0, num = 0;
    while((ch = fgetc(cipherFile)))
    {
        if (ch == EOF)
        {
            message[index] = num;
            break;
        }
        if (ch == ',')
        {
            message[index] = num;
            index++;
            num = 0;
            continue;
        }
        num *= 10;
        num += atoi(&ch);
    }
    fclose(cipherFile);

    // Get sum from decode function
    int sum = decode(message, index+1);

    printf("\nSum of decoded ASCII Chars: %i\n", sum);
    return 0;
}
